File: 1367-linked-list-in-binary-tree/1367-linked-list-in-binary-tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def check(self, head: Optional[ListNode], root: Optional[TreeNode]) -> bool:
        if head is None:
            return True
        if root is None:
            return False
        if head.val == root.val:
            return self.check(head.next, root.left) or self.check(head.next, root.right)
        return False

    def isSubPath(self, head: Optional[ListNode], root: Optional[TreeNode]) -> bool:
        if root is None:
            return False
        if head.val == root.val and self.check(head, root):
            return True
        return self.isSubPath(head, root.left) or self.isSubPath(head, root.right)

File: LeetCode/1367-linked-list-in-binary-tree/1367-linked-list-in-binary-tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool check(ListNode* head, TreeNode* root) {
        if (head == nullptr) return true;
        if (root == nullptr) return false;
        
        if (head->val == root->val) {
            return check(head->next, root->left) || check(head->next, root->right);
        }
        return false;
    }

    bool solve(ListNode* head, TreeNode* root) {
        if (root == nullptr) return false;
        
        if (head->val == root->val && check(head, root)) {
            return true;
        }
        
        return solve(head, root->left) || solve(head, root->right);
    }

    bool isSubPath(ListNode* head, TreeNode* root) {
        return solve(head, root);
    }
};